bson/_cbsonmodule.c

```diff
@@ -501,17 +527,35 @@ void destroy_codec_options(codec_options_t* options) {
     Py_CLEAR(options->document_class);
     Py_CLEAR(options->tzinfo);
     Py_CLEAR(options->options_obj);
+    Py_CLEAR(options->type_registry.registry_obj);
 }
 
 static int write_element_to_buffer(PyObject* self, buffer_t buffer,
                                    int type_byte, PyObject* value,
                                    unsigned char check_keys,
                                    const codec_options_t* options) {
     int result;
+    PyObject* value_type = NULL;
+    PyObject* converter = NULL;
+
+    if ((value_type = PyObject_Type(value)) == NULL)
+        return 0;
+
     if(Py_EnterRecursiveCall(" while encoding an object to BSON "))
         return 0;
-    result = _write_element_to_buffer(self, buffer, type_byte,
-                                      value, check_keys, options);
+
+    if ((converter = PyDict_GetItem(options->type_registry.encoder_map, value_type)) != NULL) {
+        /* Transform types that have a registered converter.
+         * A new reference is created upon transformation. */
+        PyObject* new_value = PyObject_CallFunctionObjArgs(converter, value, NULL);
```

Comment:
Need to check the return value for NULL and add error handling to decref `value_type`. I suggest rewriting to decref `value_type` ASAP, like this:
```
converter = PyDict_GetItem(options->type_registry.encoder_map, value_type)
Py_DECREF(value_type);
if (converter != NULL) {....
```